content/posts/a_confusing_build_failure.md

```diff
@@ -0,0 +1,122 @@
+
+At last, the mystery has been solved but let's dig a little deeper. Why did a test try to mock an otherwise concrete class? And when/why was the mockito-inline dependency introduced first?
+
+The Kotlin class was mocked just as a filler object for the constructor. No verifications were performed on that mocked object - neither was it used to stub a value. Hence, the concrete class should be used as is or in case of an interface, which cannot be instantiated, a dummy test double. If a concrete class was mocked because it was difficult to instantiate(for example because it has too many dependencies on its constructor) that is a moment where it is needed to pause and
+think. As for value(`data`) classes? 
+
+Quoted directly from the book *Growing object-oriented Guided by Tests* [ [^1] ]:
+
+> ***Don’t Mock Values***
+
+On the other hand, mockito-inline was introduced in the submodule in order to verify the behavior of a third-party library. Let's say that it was needed to test the persistence layer of an app. We could either run an integrated test in order to check if the code we've written works well against the real database, or we could mock the database as below.
+
+```java
+    @Test
+    public void givenDbIsEmpty_whenInsertData_thenDbSavesTheNewEntry() throws IOException {
+        /* DAO: Data Access Object
+         * usually referred to an object that fetches 
+         * data from a database or other kinds of data source. */
+        SomeDAO aDao = new SomeDAO();
+        database = mock(ExternalDatabase.class); 
+        when(database.select(any())).thenReturn(new SomeDAO());
+
+        database.insert(aDao);
+
+        SomeDAO savedDao = database.select(aDao);
+        assertThat(selectedDao, is(equalTo(aDao);
+    }
+```
+
+First of all, the problem with this test is that it actually tests nothing. The class under test is mocked and its result is stubbed, so it seems all that is tested here is the ability of the mock framework to stub a value. These kinds of tests give us no feedback and no confidence, in other words they provide zero value. We should instead automate the test for the real thing or test it manually.
+
+What's more, we are not listening to the tests. Instead of trying to guess how the library should respond to a request and find ways to mock/stub whatever is required to feed that answer to an assertion, we should add an Adapter layer over the third-party code, test the Adapter with the library and mock callbacks that we may pass to the Adapter layer. 
+
+Perhaps some direct quotes can paint the picture in a better way [ [^2] ]:
+
+> The mess in such tests is telling us that the design isn’t right but, instead of fixing the problem by improving the code, we have to carry the extra complexity in both code and test.
+
+> A second risk is that we have to be sure that the behavior we stub or mock matches what the external library will actually do.
+
+## Summing up
```

Comment:
Perhaps enhance the title to "Summing up - Guidelines" ? This section is something more than just a rehash of what you have already described.

Reply:
Split up in two sections, first one _Mitigate the many faces of the problem_ and then _Summing up_